### PR TITLE
Support Python Enums as values for Enum trait

### PR DESCRIPTION
**Pull request by corranwebster**
A fairly straightforward change.  Given:
```
class FooEnum(enum.Enum):
     foo = 0
     bar = 1
     baz = 2
```
you can do
```
class Test(HasTraits):
    value = Enum(FooEnum)
    value_default = Enum(FooEnum.bar, FooEnum)
```

Using Python `Enum` with traits `Enum`s can replace the many of the needs for the use of mapped `Trait`s since the enums hold both a value and a name out of the box.

Edit:

More work on this, with a couple of drive-by fixes:

- debug issue with editor; turns out to be a bug in the base `EditorWithList` class
- allow sets to be used in the `Enum` trait as well.
- use `xgetattr` instead of `eval` for dynamic values

### CODE REVIEW

**Comment by corranwebster** [comment]
The TraitsUI piece isn't working correctly right now.

**Comment by corranwebster** [comment]
The issue with the editor is that https://github.com/enthought/traitsui/blob/master/traitsui/editor_factory.py#L317 should be listening to `format_func` and `format_str` as well, so that after init is done the mappings are all set up correctly with the _final_ state.  Otherwise whether or not you get the correct value depends on trait initialization order.

File: traits/trait_types.py [hunk 1]
```diff
@@ -2020,25 +2037,24 @@ def create_editor(self):
...
         """ Returns the current value of a dynamic enum trait.
         """
         value = self.get_value(object, name, trait)
-        values = eval(self.values)
+        values = xgetattr(object, self.name)
         if value not in values:
-            value = None
-            if len(values) > 0:
-                value = values[0]
+            value = collection_default(values)
```
Comment (kitchoi) [inline traits/trait_types.py]: I was trying to tackle #389 and this line is all I needed to solve #389.

**Comment by mdickinson** [comment]
I get an error running the test suite under Python 3.8:
```
======================================================================
ERROR: test_enum_enum (traits.tests.test_enum.EnumTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/mdickinson/Enthought/ETS/traits/traits/tests/test_enum.py", line 129, in test_enum_enum
    self.assertEqual(example.value_name, FooEnum.foo)
  File "/Users/mdickinson/Enthought/ETS/traits/traits/trait_types.py", line 2049, in _get
    if value not in values:
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/enum.py", line 310, in __contains__
    raise TypeError(
TypeError: unsupported operand type(s) for 'in': '_Undefined' and 'EnumMeta'
```

**Comment by mdickinson** [comment]
Related: https://bugs.python.org/issue33217. I'm not sure why it was considered a good idea to raise `TypeError` rather than simply returning `False`.

File: traits/trait_base.py [hunk 2]
```diff
@@ -194,6 +197,23 @@ def strx(arg):
...
+
+    Parameters
+    ----------
+    collection : collection
+        A Python collection, which is presumed to be repeatably iterable.
+
+    Returns
+    -------
+    default : any
+        The first item of the collectio, or None if the collection is empty.
```
Comment (mdickinson) [inline traits/trait_base.py]: Typo: "collectio"

File: traits/trait_base.py [hunk 3]
```diff
@@ -33,6 +34,8 @@
 
 SequenceTypes = (list, tuple)
 
+EnumTypes = (list, tuple, set, enum.EnumMeta)
```
Comment (mdickinson) [inline traits/trait_base.py]: `enum.EnumMeta` doesn't seen right here: an instance of `enum.EnumMeta` is an enumeration _class_, where presumably we want to check for an instance of an enumeration class. Do you just want `enum.Enum` here?

Comment (mdickinson) [reply traits/trait_base.py]: > Do you just want `enum.Enum` here?

No, you don't. I misunderstood how this was being used. Closing.

File: traits/trait_base.py [hunk 4]
```diff
@@ -33,6 +34,8 @@
...
+
```
Comment (mdickinson) [inline traits/trait_base.py]: Supporting `set` here is new, right? I'm not convinced we want to do that: in particular, getting unpredictable defaults could be a problem.

File: traits/trait_types.py (hunk 1 above)
Comment (mdickinson) [reply traits/trait_types.py]: @kitchoi can you clarify: if this PR gets merged, does that mean #389 can be closed? Or is there more work to do between merging this PR and solving #389?

Comment (kitchoi) [reply traits/trait_types.py]: I think #389 can be closed by this PR (as of now) without more being done to it. 

While #389 concerns the dictionary only, the idea is to relax the requirement on the `values` supporting `__index__`.  As long as `values` implements `__iter__`, Enum can obtain the allowed values. 

I understand your concern with the potential non-deterministic behaviour, e.g. for `set` and the `dict` prior to Python 3.6.  But if, say, `values` is an instance of `OrderedDict`, the `Enum` on master will fail trying to use the first key as a default.  One has to make another property traits to return a list from the keys of `OrderedDict`.  I guess it is about how opinionated this `Enum` should be when it comes to setting a default value.

File: traits/trait_base.py (hunk 4 above)
Comment (corranwebster) [reply traits/trait_base.py]: Ah - I had assumed that set had picked up the ordering behaviour of dicts.  I will remove this.

File: traits/trait_types.py (hunk 1 above)
Comment (kitchoi) [reply traits/trait_types.py]: To verify:
```
from traits.api import HasTraits, Enum, Any


class Foo(HasTraits):

    a = Enum(values="b")

    b = Any()

    def _b_default(self):
        return {"a": 1, "b": 2}


if __name__ == "__main__":
    f = Foo()
    # f.configure_traits()
    print(repr(f.a))
```

On this branch, a value `'a'` was printed.
On master, this results in an error:
```
Traceback (most recent call last):
  File "dict_enum.py", line 17, in <module>
    print(repr(f.a))
  File "/Users/kchoi/ETS/traits/traits/trait_types.py", line 2030, in _get
    value = values[0]
KeyError: 0
```

**Comment by corranwebster** [comment]
I don't have an easy way to check on Python 3.8, but I have attempted to solve the `value in values` issue by defining a `safe_contains` function that does the containment check, but catches `TypeError` and returns `False` in that case.  This hopefully will do something reasonable in the test case.

**Comment by mdickinson** [comment]
> I don't have an easy way to check on Python 3.8 [...]

Confirmed working on 3.8.

File: traits/trait_types.py (hunk 1 above)
Comment (mdickinson) [reply traits/trait_types.py]: Hmm. I think I liked the error better. :-)

**Pull request merged**